File: zig/src/bplustree.zig
const std = @import("std");

pub fn BPlusTree(comptime K: type, comptime V: type) type {
    return struct {
        const Self = @This();
        const KeyType = K;
        const ValueType = V;
        
        pub const Error = error{
            KeyNotFound,
            OutOfMemory,
        };
        
        pub const Entry = struct {
            key: KeyType,
            value: ValueType,
        };
        
        pub const Iterator = struct {
            current_node: ?*Node,
            current_index: usize,
            
            pub fn next(self: *Iterator) ?Entry {
                if (self.current_node == null) return null;
                
                const node = self.current_node.?;
                
                // Check if we have more items in current node
                if (self.current_index < node.keys.items.len) {
                    const entry = Entry{
                        .key = node.keys.items[self.current_index],
                        .value = node.values.?.items[self.current_index],
                    };
                    self.current_index += 1;
                    return entry;
                }
                
                // Move to next leaf node
                self.current_node = node.next;
                self.current_index = 0;
                
                // Try again with next node
                return self.next();
            }
        };
        
        pub const ReverseIterator = struct {
            current_node: ?*Node,
            current_index: usize,
            
            pub fn next(self: *ReverseIterator) ?Entry {
                if (self.current_node == null) return null;
                
                const node = self.current_node.?;
                
                // Check if we have more items in current node
                if (self.current_index > 0) {
                    self.current_index -= 1;
                    const entry = Entry{
                        .key = node.keys.items[self.current_index],
                        .value = node.values.?.items[self.current_index],
                    };
                    return entry;
                }
                
                // Move to previous leaf node
                self.current_node = node.prev;
                if (self.current_node) |prev_node| {
                    self.current_index = prev_node.keys.items.len;
                }
                
                // Try again with previous node
                return self.next();
            }
        };
        
        const NodeType = enum { leaf, branch };
        
        const Node = struct {
            node_type: NodeType,
            keys: std.ArrayList(KeyType),
            
            // For leaf nodes
            values: ?std.ArrayList(ValueType),
            next: ?*Node,  // Link to next leaf
            prev: ?*Node,  // Link to previous leaf
            
            // For branch nodes
            children: ?std.ArrayList(*Node),
            
            fn initLeaf(allocator: std.mem.Allocator) !*Node {
                const node = try allocator.create(Node);
                node.* = Node{
                    .node_type = .leaf,
                    .keys = std.ArrayList(KeyType).init(allocator),
                    .values = std.ArrayList(ValueType).init(allocator),
                    .children = null,
                    .next = null,
                    .prev = null,
                };
                return node;
            }
            
            fn initBranch(allocator: std.mem.Allocator) !*Node {
                const node = try allocator.create(Node);
                node.* = Node{
                    .node_type = .branch,
                    .keys = std.ArrayList(KeyType).init(allocator),
                    .values = null,
                    .children = std.ArrayList(*Node).init(allocator),
                    .next = null,
                    .prev = null,
                };
                return node;
            }
            
            fn deinit(self: *Node, allocator: std.mem.Allocator) void {
                self.keys.deinit();
                if (self.values) |*values| {
                    values.deinit();
                }
                if (self.children) |*children| {
                    for (children.items) |child| {
                        child.deinit(allocator);
                    }
                    children.deinit();
                }
                allocator.destroy(self);
            }
            
            fn isFull(self: *const Node, capacity: usize) bool {
                return self.keys.items.len >= capacity;
            }
        };
        
        allocator: std.mem.Allocator,
        capacity: usize,
        size: usize,
        root: ?*Node,
        
        pub fn init(allocator: std.mem.Allocator, capacity: usize) Self {
            return Self{
                .allocator = allocator,
                .capacity = capacity,
                .size = 0,
                .root = null,
            };
        }
        
        pub fn deinit(self: *Self) void {
            if (self.root) |root| {
                root.deinit(self.allocator);
            }
        }
        
        pub fn len(self: *const Self) usize {
            return self.size;
        }
        
        pub fn insert(self: *Self, key: KeyType, value: ValueType) !void {
            if (self.root == null) {
                self.root = try Node.initLeaf(self.allocator);
            }
            
            // If root is full and is a leaf, split it
            if (self.root.?.node_type == .leaf and self.root.?.isFull(self.capacity)) {
                const old_root = self.root.?;
                const new_leaf = try self.splitLeaf(old_root);
                
                // Create new root
                var new_root = try Node.initBranch(self.allocator);
                try new_root.keys.append(new_leaf.keys.items[0]);
                try new_root.children.?.append(old_root);
                try new_root.children.?.append(new_leaf);
                
                self.root = new_root;
            }
            
            // Now insert normally
            try self.insertIntoNode(self.root.?, key, value);
            self.size += 1;
        }
        
        fn insertIntoNode(self: *Self, node: *Node, key: KeyType, value: ValueType) !void {
            if (node.node_type == .leaf) {
                // Find the position to insert
                var insert_pos: usize = 0;
                var found = false;
                
                for (node.keys.items, 0..) |k, i| {
                    if (k == key) {
                        // Key already exists, update value
                        node.values.?.items[i] = value;
                        self.size -= 1; // Compensate for the increment in insert()
                        return;
                    }
                    if (k > key) {
                        insert_pos = i;
                        found = true;
                        break;
                    }
                }
                
                if (!found) {
                    insert_pos = node.keys.items.len;
                }
                
                // Insert at the correct position
                try node.keys.insert(insert_pos, key);
                try node.values.?.insert(insert_pos, value);
            } else {
                // Find child to insert into
                var child_idx: usize = 0;
                for (node.keys.items, 0..) |k, i| {
                    if (key < k) {
                        break;
                    }
                    child_idx = i + 1;
                }
                if (child_idx >= node.children.?.items.len) {
                    child_idx = node.children.?.items.len - 1;
                }
                
                const child = node.children.?.items[child_idx];
                
                // Check if child needs splitting before insertion
                if (child.isFull(self.capacity)) {
                    // Split the child
                    if (child.node_type == .leaf) {
                        const new_node = try self.splitLeaf(child);
                        
                        // Insert the new key into parent
                        const split_key = new_node.keys.items[0];
                        const insert_pos: usize = child_idx + 1;
                        
                        // Insert new key and child pointer
                        try node.keys.insert(child_idx, split_key);
                        try node.children.?.insert(insert_pos, new_node);
                        
                        // Decide which child to insert into
                        if (key >= split_key) {
                            try self.insertIntoNode(new_node, key, value);
                        } else {
                            try self.insertIntoNode(child, key, value);
                        }
                    }
                } else {
                    try self.insertIntoNode(child, key, value);
                }
            }
        }
        
        pub fn get(self: *const Self, key: KeyType) ?ValueType {
            if (self.root == null) return null;
            
            var current = self.root.?;
            
            // Navigate down the tree
            while (current.node_type == .branch) {
                var child_idx: usize = 0;
                for (current.keys.items, 0..) |k, i| {
                    if (key < k) {
                        break;
                    }
                    child_idx = i + 1;
                }
                if (child_idx >= current.children.?.items.len) {
                    child_idx = current.children.?.items.len - 1;
                }
                current = current.children.?.items[child_idx];
            }
            
            // Search in leaf node
            for (current.keys.items, 0..) |k, i| {
                if (k == key) {
                    return current.values.?.items[i];
                }
            }
            return null;
        }
        
        pub fn getHeight(self: *const Self) usize {
            if (self.root == null) return 0;
            
            var height: usize = 1;
            var current = self.root.?;
            
            while (current.node_type == .branch and current.children.?.items.len > 0) {
                height += 1;
                current = current.children.?.items[0];
            }
            
            return height;
        }
        
        fn splitLeaf(self: *Self, leaf: *Node) !*Node {
            const mid = leaf.keys.items.len / 2;
            
            // Create new leaf
            var new_leaf = try Node.initLeaf(self.allocator);
            
            // Move half the keys and values to new leaf
            for (mid..leaf.keys.items.len) |i| {
                try new_leaf.keys.append(leaf.keys.items[i]);
                try new_leaf.values.?.append(leaf.values.?.items[i]);
            }
            
            // Remove moved items from original leaf
            leaf.keys.shrinkRetainingCapacity(mid);
            leaf.values.?.shrinkRetainingCapacity(mid);
            
            // Update leaf links
            new_leaf.next = leaf.next;
            new_leaf.prev = leaf;
            if (leaf.next) |next| {
                next.prev = new_leaf;
            }
            leaf.next = new_leaf;
            
            return new_leaf;
        }
        
        pub fn remove(self: *Self, key: KeyType) !ValueType {
            if (self.root == null) {
                return Error.KeyNotFound;
            }
            
            // Remove from the tree starting at root
            const result = try self.removeFromNode(self.root.?, key);
            
            // If root is now empty and is a branch node, promote its only child
            if (self.root.?.node_type == .branch and self.root.?.keys.items.len == 0) {
                if (self.root.?.children.?.items.len > 0) {
                    const old_root = self.root.?;
                    self.root = self.root.?.children.?.items[0];
                    
                    // Clean up old root
                    old_root.keys.deinit();
                    old_root.children.?.deinit();
                    self.allocator.destroy(old_root);
                }
            }
            
            // If root is a leaf and empty, tree is now empty
            if (self.root.?.node_type == .leaf and self.root.?.keys.items.len == 0) {
                self.root.?.deinit(self.allocator);
                self.root = null;
            }
            
            self.size -= 1;
            return result;
        }
        
        fn removeFromNode(self: *Self, node: *Node, key: KeyType) !ValueType {
            if (node.node_type == .leaf) {
                // Find and remove the key from leaf
                for (node.keys.items, 0..) |k, i| {
                    if (k == key) {
                        const value = node.values.?.items[i];
                        
                        // Remove key and value
                        _ = node.keys.orderedRemove(i);
                        _ = node.values.?.orderedRemove(i);
                        
                        return value;
                    }
                }
                return Error.KeyNotFound;
            } else {
                // Find child that contains the key
                var child_idx: usize = 0;
                for (node.keys.items, 0..) |k, i| {
                    if (key < k) {
                        break;
                    }
                    child_idx = i + 1;
                }
                if (child_idx >= node.children.?.items.len) {
                    child_idx = node.children.?.items.len - 1;
                }
                
                var child = node.children.?.items[child_idx];
                const min_keys = (self.capacity + 1) / 2 - 1;
                
                // Check if child will underflow after deletion
                if (child.keys.items.len <= min_keys) {
                    // Try to borrow from siblings or merge
                    const new_child_idx = try self.handleUnderflow(node, child_idx);
                    // Update child reference after potential merge
                    if (new_child_idx < node.children.?.items.len) {
                        child = node.children.?.items[new_child_idx];
                        child_idx = new_child_idx;
                    }
                }
                
                // Now proceed with deletion
                return try self.removeFromNode(child, key);
            }
        }
        
        fn handleUnderflow(self: *Self, parent: *Node, child_idx: usize) !usize {
            const child = parent.children.?.items[child_idx];
            const min_keys = (self.capacity + 1) / 2 - 1;
            
            // Try borrowing from left sibling
            if (child_idx > 0) {
                const left_sibling = parent.children.?.items[child_idx - 1];
                if (left_sibling.keys.items.len > min_keys) {
                    // Borrow from left sibling
                    if (child.node_type == .leaf) {
                        // Move last key from left sibling to child
                        const last_idx = left_sibling.keys.items.len - 1;
                        const borrowed_key = left_sibling.keys.items[last_idx];
                        const borrowed_value = left_sibling.values.?.items[last_idx];
                        
                        _ = left_sibling.keys.pop();
                        _ = left_sibling.values.?.pop();
                        
                        try child.keys.insert(0, borrowed_key);
                        try child.values.?.insert(0, borrowed_value);
                        
                        // Update parent key
                        parent.keys.items[child_idx - 1] = child.keys.items[0];
                    }
                    return child_idx;
                }
            }
            
            // Try borrowing from right sibling
            if (child_idx < parent.children.?.items.len - 1) {
                const right_sibling = parent.children.?.items[child_idx + 1];
                if (right_sibling.keys.items.len > min_keys) {
                    // Borrow from right sibling
                    if (child.node_type == .leaf) {
                        // Move first key from right sibling to child
                        const borrowed_key = right_sibling.keys.orderedRemove(0);
                        const borrowed_value = right_sibling.values.?.orderedRemove(0);
                        try child.keys.append(borrowed_key);
                        try child.values.?.append(borrowed_value);
                        
                        // Update parent key
                        parent.keys.items[child_idx] = right_sibling.keys.items[0];
                    }
                    return child_idx;
                }
            }
            
            // If can't borrow, merge with a sibling
            if (child_idx > 0) {
                // Merge with left sibling
                try self.mergeNodes(parent, child_idx - 1, child_idx);
                return child_idx - 1;
            } else {
                // Merge with right sibling
                try self.mergeNodes(parent, child_idx, child_idx + 1);
                return child_idx;
            }
        }
        
        fn mergeNodes(self: *Self, parent: *Node, left_idx: usize, right_idx: usize) !void {
            const left = parent.children.?.items[left_idx];
            const right = parent.children.?.items[right_idx];
            
            if (left.node_type == .leaf) {
                // Merge leaf nodes
                for (right.keys.items, 0..) |key, i| {
                    try left.keys.append(key);
                    try left.values.?.append(right.values.?.items[i]);
                }
                
                // Update leaf links
                left.next = right.next;
                if (right.next) |next| {
                    next.prev = left;
                }
                
                // Remove the merged node from parent
                _ = parent.keys.orderedRemove(left_idx);
                _ = parent.children.?.orderedRemove(right_idx);
                
                // Clean up merged node
                right.deinit(self.allocator);
            }
        }
        
        pub fn range(self: *const Self, start: KeyType, end: KeyType, results: *std.ArrayList(Entry)) !void {
            if (self.root == null) {
                return;
            }
            
            // Find the leaf containing the start key
            var current = self.root.?;
            
            // Navigate to the leaf that would contain start
            while (current.node_type == .branch) {
                var child_idx: usize = 0;
                for (current.keys.items, 0..) |k, i| {
                    if (start < k) {
                        break;
                    }
                    child_idx = i + 1;
                }
                if (child_idx >= current.children.?.items.len) {
                    child_idx = current.children.?.items.len - 1;
                }
                current = current.children.?.items[child_idx];
            }
            
            // Now traverse through leaf nodes collecting values in range
            var leaf: ?*Node = current;
            while (leaf != null) {
                const node = leaf.?;
                
                // Process keys in this leaf
                for (node.keys.items, 0..) |k, i| {
                    if (k > end) {
                        // We've passed the end of the range
                        return;
                    }
                    if (k >= start) {
                        // Key is in range
                        try results.append(.{
                            .key = k,
                            .value = node.values.?.items[i],
                        });
                    }
                }
                
                // Move to next leaf
                leaf = node.next;
            }
        }
        
        pub fn iterator(self: *const Self) Iterator {
            if (self.root == null) {
                return Iterator{
                    .current_node = null,
                    .current_index = 0,
                };
            }
            
            // Find leftmost leaf
            var current = self.root.?;
            while (current.node_type == .branch) {
                current = current.children.?.items[0];
            }
            
            return Iterator{
                .current_node = current,
                .current_index = 0,
            };
        }
        
        pub fn reverseIterator(self: *const Self) ReverseIterator {
            if (self.root == null) {
                return ReverseIterator{
                    .current_node = null,
                    .current_index = 0,
                };
            }
            
            // Find rightmost leaf
            var current = self.root.?;
            while (current.node_type == .branch) {
                current = current.children.?.items[current.children.?.items.len - 1];
            }
            
            return ReverseIterator{
                .current_node = current,
                .current_index = current.keys.items.len,
            };
        }
        
        pub fn contains(self: *const Self, key: KeyType) bool {
            return self.get(key) != null;
        }
        
        pub fn clear(self: *Self) void {
            if (self.root) |root| {
                root.deinit(self.allocator);
                self.root = null;
            }
            self.size = 0;
        }
    };
}

File: go/internal/bplustree/bplustree.go
package bplustree

import (
	"cmp"
	"errors"
)

// ErrKeyNotFound is returned when trying to delete a non-existent key
var ErrKeyNotFound = errors.New("key not found")

// Entry represents a key-value pair
type Entry[K cmp.Ordered, V any] struct {
	Key   K
	Value V
}

// nodeType represents the type of a B+ tree node
type nodeType int

const (
	leafNode nodeType = iota
	branchNode
)

// node represents a B+ tree node
type node[K cmp.Ordered, V any] struct {
	nodeType nodeType
	keys     []K
	
	// For leaf nodes
	values []V
	next   *node[K, V]
	prev   *node[K, V]
	
	// For branch nodes
	children []*node[K, V]
}

// BPlusTree represents a B+ tree data structure
type BPlusTree[K cmp.Ordered, V any] struct {
	root     *node[K, V]
	capacity int
	size     int
}

// New creates a new B+ tree with the given capacity
func New[K cmp.Ordered, V any](capacity int) *BPlusTree[K, V] {
	if capacity < 3 {
		capacity = 3
	}
	return &BPlusTree[K, V]{
		capacity: capacity,
		size:     0,
		root:     nil,
	}
}

// Len returns the number of key-value pairs in the tree
func (t *BPlusTree[K, V]) Len() int {
	return t.size
}

// Height returns the height of the tree
func (t *BPlusTree[K, V]) Height() int {
	if t.root == nil {
		return 0
	}
	
	height := 1
	current := t.root
	for current.nodeType == branchNode && len(current.children) > 0 {
		height++
		current = current.children[0]
	}
	return height
}

// Insert inserts or updates a key-value pair in the tree
func (t *BPlusTree[K, V]) Insert(key K, value V) {
	if t.root == nil {
		t.root = &node[K, V]{
			nodeType: leafNode,
			keys:     []K{key},
			values:   []V{value},
		}
		t.size++
		return
	}
	
	// Check if root needs splitting
	if t.root.nodeType == leafNode && len(t.root.keys) >= t.capacity {
		// Split root
		newNode := t.splitLeaf(t.root)
		
		// Create new root
		newRoot := &node[K, V]{
			nodeType: branchNode,
			keys:     []K{newNode.keys[0]},
			children: []*node[K, V]{t.root, newNode},
		}
		t.root = newRoot
	}
	
	// Insert into the tree
	t.insertIntoNode(t.root, key, value)
}

// insertIntoNode inserts a key-value pair into the given node
func (t *BPlusTree[K, V]) insertIntoNode(n *node[K, V], key K, value V) {
	if n.nodeType == leafNode {
		// Find position to insert
		pos := 0
		for pos < len(n.keys) && n.keys[pos] < key {
			pos++
		}
		
		// Check if key already exists
		if pos < len(n.keys) && n.keys[pos] == key {
			n.values[pos] = value
			return
		}
		
		// Insert at position
		n.keys = append(n.keys[:pos], append([]K{key}, n.keys[pos:]...)...)
		n.values = append(n.values[:pos], append([]V{value}, n.values[pos:]...)...)
		t.size++
	} else {
		// Find child to insert into
		childIdx := 0
		for childIdx < len(n.keys) && key >= n.keys[childIdx] {
			childIdx++
		}
		if childIdx >= len(n.children) {
			childIdx = len(n.children) - 1
		}
		
		child := n.children[childIdx]
		
		// Check if child needs splitting
		if len(child.keys) >= t.capacity {
			var newNode *node[K, V]
			if child.nodeType == leafNode {
				newNode = t.splitLeaf(child)
			} else {
				newNode = t.splitBranch(child)
			}
			
			// Insert new key into parent
			splitKey := newNode.keys[0]
			
			// Insert key and child pointer
			n.keys = append(n.keys[:childIdx], append([]K{splitKey}, n.keys[childIdx:]...)...)
			n.children = append(n.children[:childIdx+1], append([]*node[K, V]{newNode}, n.children[childIdx+1:]...)...)
			
			// Decide which child to insert into
			if key >= splitKey {
				t.insertIntoNode(newNode, key, value)
			} else {
				t.insertIntoNode(child, key, value)
			}
		} else {
			t.insertIntoNode(child, key, value)
		}
	}
}

// splitLeaf splits a leaf node and returns the new node
func (t *BPlusTree[K, V]) splitLeaf(leaf *node[K, V]) *node[K, V] {
	mid := len(leaf.keys) / 2
	
	// Create new leaf with right half
	newLeaf := &node[K, V]{
		nodeType: leafNode,
		keys:     append([]K(nil), leaf.keys[mid:]...),
		values:   append([]V(nil), leaf.values[mid:]...),
		next:     leaf.next,
		prev:     leaf,
	}
	
	// Update leaf links
	if leaf.next != nil {
		leaf.next.prev = newLeaf
	}
	leaf.next = newLeaf
	
	// Keep only left half in original leaf
	leaf.keys = leaf.keys[:mid]
	leaf.values = leaf.values[:mid]
	
	return newLeaf
}

// splitBranch splits a branch node and returns the new node
func (t *BPlusTree[K, V]) splitBranch(branch *node[K, V]) *node[K, V] {
	mid := len(branch.keys) / 2
	
	// Create new branch with right half
	newBranch := &node[K, V]{
		nodeType: branchNode,
		keys:     append([]K(nil), branch.keys[mid+1:]...),
		children: append([]*node[K, V](nil), branch.children[mid+1:]...),
	}
	
	// Keep only left half in original branch
	branch.keys = branch.keys[:mid]
	branch.children = branch.children[:mid+1]
	
	return newBranch
}

// Get retrieves the value associated with a key
func (t *BPlusTree[K, V]) Get(key K) (V, bool) {
	var zero V
	if t.root == nil {
		return zero, false
	}
	
	current := t.root
	
	// Navigate to leaf
	for current.nodeType == branchNode {
		childIdx := 0
		for childIdx < len(current.keys) && key >= current.keys[childIdx] {
			childIdx++
		}
		if childIdx >= len(current.children) {
			childIdx = len(current.children) - 1
		}
		current = current.children[childIdx]
	}
	
	// Search in leaf
	for i, k := range current.keys {
		if k == key {
			return current.values[i], true
		}
	}
	
	return zero, false
}

// Delete removes a key-value pair from the tree
func (t *BPlusTree[K, V]) Delete(key K) (V, error) {
	var zero V
	if t.root == nil {
		return zero, ErrKeyNotFound
	}
	
	// Delete from the tree
	value, err := t.deleteFromNode(t.root, key)
	if err != nil {
		return zero, err
	}
	
	// Handle root becoming empty
	if t.root.nodeType == branchNode && len(t.root.keys) == 0 {
		if len(t.root.children) > 0 {
			t.root = t.root.children[0]
		}
	}
	
	// Handle empty tree
	if t.root.nodeType == leafNode && len(t.root.keys) == 0 {
		t.root = nil
	}
	
	t.size--
	return value, nil
}

// deleteFromNode deletes a key from the given node
func (t *BPlusTree[K, V]) deleteFromNode(n *node[K, V], key K) (V, error) {
	var zero V
	
	if n.nodeType == leafNode {
		// Find and delete from leaf
		for i, k := range n.keys {
			if k == key {
				value := n.values[i]
				
				// Remove key and value
				n.keys = append(n.keys[:i], n.keys[i+1:]...)
				n.values = append(n.values[:i], n.values[i+1:]...)
				
				return value, nil
			}
		}
		return zero, ErrKeyNotFound
	}
	
	// Find child containing the key
	childIdx := 0
	for childIdx < len(n.keys) && key >= n.keys[childIdx] {
		childIdx++
	}
	if childIdx >= len(n.children) {
		childIdx = len(n.children) - 1
	}
	
	child := n.children[childIdx]
	minKeys := (t.capacity + 1) / 2 - 1
	
	// Check if child will underflow
	if len(child.keys) <= minKeys {
		// Try to handle underflow
		t.handleUnderflow(n, childIdx)
		// Re-find child after potential merge
		if childIdx >= len(n.children) {
			childIdx = len(n.children) - 1
		}
		child = n.children[childIdx]
	}
	
	return t.deleteFromNode(child, key)
}

// handleUnderflow handles node underflow by redistribution or merging
func (t *BPlusTree[K, V]) handleUnderflow(parent *node[K, V], childIdx int) {
	child := parent.children[childIdx]
	minKeys := (t.capacity + 1) / 2 - 1
	
	// Try borrowing from left sibling
	if childIdx > 0 {
		leftSibling := parent.children[childIdx-1]
		if len(leftSibling.keys) > minKeys {
			if child.nodeType == leafNode {
				// Borrow from left sibling
				borrowedKey := leftSibling.keys[len(leftSibling.keys)-1]
				borrowedValue := leftSibling.values[len(leftSibling.values)-1]
				
				leftSibling.keys = leftSibling.keys[:len(leftSibling.keys)-1]
				leftSibling.values = leftSibling.values[:len(leftSibling.values)-1]
				
				child.keys = append([]K{borrowedKey}, child.keys...)
				child.values = append([]V{borrowedValue}, child.values...)
				
				parent.keys[childIdx-1] = child.keys[0]
			}
			return
		}
	}
	
	// Try borrowing from right sibling
	if childIdx < len(parent.children)-1 {
		rightSibling := parent.children[childIdx+1]
		if len(rightSibling.keys) > minKeys {
			if child.nodeType == leafNode {
				// Borrow from right sibling
				borrowedKey := rightSibling.keys[0]
				borrowedValue := rightSibling.values[0]
				
				rightSibling.keys = rightSibling.keys[1:]
				rightSibling.values = rightSibling.values[1:]
				
				child.keys = append(child.keys, borrowedKey)
				child.values = append(child.values, borrowedValue)
				
				parent.keys[childIdx] = rightSibling.keys[0]
			}
			return
		}
	}
	
	// Merge with sibling
	if childIdx > 0 {
		// Merge with left sibling
		t.mergeNodes(parent, childIdx-1, childIdx)
	} else {
		// Merge with right sibling
		t.mergeNodes(parent, childIdx, childIdx+1)
	}
}

// mergeNodes merges two sibling nodes
func (t *BPlusTree[K, V]) mergeNodes(parent *node[K, V], leftIdx, rightIdx int) {
	left := parent.children[leftIdx]
	right := parent.children[rightIdx]
	
	if left.nodeType == leafNode {
		// Merge leaf nodes
		left.keys = append(left.keys, right.keys...)
		left.values = append(left.values, right.values...)
		
		// Update leaf links
		left.next = right.next
		if right.next != nil {
			right.next.prev = left
		}
		
		// Remove from parent
		parent.keys = append(parent.keys[:leftIdx], parent.keys[leftIdx+1:]...)
		parent.children = append(parent.children[:rightIdx], parent.children[rightIdx+1:]...)
	}
}

// Range returns all entries with keys in the range [start, end]
func (t *BPlusTree[K, V]) Range(start, end K) []Entry[K, V] {
	var results []Entry[K, V]
	
	if t.root == nil || start > end {
		return results
	}
	
	// Find the first leaf that could contain start
	current := t.root
	for current.nodeType == branchNode {
		childIdx := 0
		for childIdx < len(current.keys) && start >= current.keys[childIdx] {
			childIdx++
		}
		if childIdx >= len(current.children) {
			childIdx = len(current.children) - 1
		}
		current = current.children[childIdx]
	}
	
	// Traverse leaves collecting entries in range
	for current != nil {
		for i, key := range current.keys {
			if key > end {
				return results
			}
			if key >= start {
				results = append(results, Entry[K, V]{
					Key:   key,
					Value: current.values[i],
				})
			}
		}
		current = current.next
	}
	
	return results
}

// Iterator represents a forward iterator over the B+ tree
type Iterator[K cmp.Ordered, V any] struct {
	currentNode  *node[K, V]
	currentIndex int
	key          K
	value        V
}

// Iterator returns a new forward iterator
func (t *BPlusTree[K, V]) Iterator() *Iterator[K, V] {
	if t.root == nil {
		return &Iterator[K, V]{
			currentNode:  nil,
			currentIndex: 0,
		}
	}
	
	// Find leftmost leaf
	current := t.root
	for current.nodeType == branchNode {
		current = current.children[0]
	}
	
	return &Iterator[K, V]{
		currentNode:  current,
		currentIndex: -1,
	}
}

// Next advances the iterator and returns true if there are more elements
func (it *Iterator[K, V]) Next() bool {
	if it.currentNode == nil {
		return false
	}
	
	it.currentIndex++
	
	// Check if we have more items in current node
	if it.currentIndex < len(it.currentNode.keys) {
		it.key = it.currentNode.keys[it.currentIndex]
		it.value = it.currentNode.values[it.currentIndex]
		return true
	}
	
	// Move to next leaf node
	it.currentNode = it.currentNode.next
	it.currentIndex = 0
	
	// Try again with next node
	if it.currentNode != nil && len(it.currentNode.keys) > 0 {
		it.key = it.currentNode.keys[0]
		it.value = it.currentNode.values[0]
		return true
	}
	
	return false
}

// Key returns the current key
func (it *Iterator[K, V]) Key() K {
	return it.key
}

// Value returns the current value
func (it *Iterator[K, V]) Value() V {
	return it.value
}

// ReverseIterator represents a reverse iterator over the B+ tree
type ReverseIterator[K cmp.Ordered, V any] struct {
	currentNode  *node[K, V]
	currentIndex int
	key          K
	value        V
}

// ReverseIterator returns a new reverse iterator
func (t *BPlusTree[K, V]) ReverseIterator() *ReverseIterator[K, V] {
	if t.root == nil {
		return &ReverseIterator[K, V]{
			currentNode:  nil,
			currentIndex: 0,
		}
	}
	
	// Find rightmost leaf
	current := t.root
	for current.nodeType == branchNode {
		current = current.children[len(current.children)-1]
	}
	
	return &ReverseIterator[K, V]{
		currentNode:  current,
		currentIndex: len(current.keys),
	}
}

// Next advances the reverse iterator and returns true if there are more elements
func (it *ReverseIterator[K, V]) Next() bool {
	if it.currentNode == nil {
		return false
	}
	
	it.currentIndex--
	
	// Check if we have more items in current node
	if it.currentIndex >= 0 {
		it.key = it.currentNode.keys[it.currentIndex]
		it.value = it.currentNode.values[it.currentIndex]
		return true
	}
	
	// Move to previous leaf node
	it.currentNode = it.currentNode.prev
	if it.currentNode != nil {
		it.currentIndex = len(it.currentNode.keys) - 1
		if it.currentIndex >= 0 {
			it.key = it.currentNode.keys[it.currentIndex]
			it.value = it.currentNode.values[it.currentIndex]
			return true
		}
	}
	
	return false
}

// Key returns the current key
func (it *ReverseIterator[K, V]) Key() K {
	return it.key
}

// Value returns the current value
func (it *ReverseIterator[K, V]) Value() V {
	return it.value
}

// Contains checks if a key exists in the tree
func (t *BPlusTree[K, V]) Contains(key K) bool {
	_, found := t.Get(key)
	return found
}

// Clear removes all elements from the tree
func (t *BPlusTree[K, V]) Clear() {
	t.root = nil
	t.size = 0
}